core/org.wso2.carbon.core/src/main/java/org/wso2/carbon/core/clustering/hazelcast/kubernetes/KubernetesMembershipScheme.java

```diff
@@ -0,0 +1,164 @@
+                                      HazelcastInstance primaryHazelcastInstance, List<ClusteringMessage> messageBuffer) {
+
+        this.parameters = parameters;
+        this.primaryHazelcastInstance = primaryHazelcastInstance;
+        this.messageBuffer = messageBuffer;
+        this.nwConfig = config.getNetworkConfig();
+    }
+
+    @Override
+    public void setPrimaryHazelcastInstance(HazelcastInstance primaryHazelcastInstance) {
+
+        this.primaryHazelcastInstance = primaryHazelcastInstance;
+    }
+
+    @Override
+    public void setLocalMember(Member localMember) {
+
+    }
+
+    @Override
+    public void setCarbonCluster(HazelcastCarbonClusterImpl hazelcastCarbonCluster) {
+
+        this.carbonCluster = hazelcastCarbonCluster;
+    }
+
+    @Override
+    public void init() {
+
+        String localMemberPort = "4000";
+        Parameter namespace = getParameter(KubernetesConstants.NAMESPACE_PROPERTY);
+        Parameter serviceName = getParameter(KubernetesConstants.SERVICE_NAME_PROPERTY);
+
+        log.info("Initializing kubernetes membership scheme...");
+        nwConfig.getJoin().getMulticastConfig().setEnabled(false);
+        nwConfig.getJoin().getAwsConfig().setEnabled(false);
+        nwConfig.getJoin().getKubernetesConfig().setEnabled(true);
+
+        Parameter localMemberPortParam = getParameter(HazelcastConstants.LOCAL_MEMBER_PORT);
+        if (localMemberPortParam != null) {
+            localMemberPort = ((String) localMemberPortParam.getValue()).trim();
+            nwConfig.getJoin().getKubernetesConfig().setProperty(KubernetesConstants.SERVICE_PORT, localMemberPort);
+        } else {
+            nwConfig.getJoin().getKubernetesConfig().setProperty(KubernetesConstants.SERVICE_PORT, localMemberPort);
+        }
+
+        if (namespace != null) {
+            nwConfig.getJoin().getKubernetesConfig().setProperty(KubernetesConstants.NAMESPACE, ((String) namespace.getValue()).trim());
+        }
+        if (serviceName != null) {
+            nwConfig.getJoin().getKubernetesConfig().setProperty(KubernetesConstants.SERVICE_NAME, ((String) serviceName.getValue()).trim());
+        }
+    }
+
+    @Override
+    public void joinGroup() {
+
+        primaryHazelcastInstance.getCluster().addMembershipListener(new KubernetesMembershipSchemeListener());
+    }
+
+    private Parameter getParameter(String name) {
+
+        return parameters.get(name);
+    }
+
+    /**
+     * Kubernetes membership scheme listener
+     */
+    private class KubernetesMembershipSchemeListener implements MembershipListener {
+
+        @Override
+        public void memberAdded(MembershipEvent membershipEvent) {
+
+            Member member = membershipEvent.getMember();
+            TcpIpConfig tcpIpConfig = nwConfig.getJoin().getTcpIpConfig();
+            List<String> memberList = tcpIpConfig.getMembers();
+            if (!memberList.contains(member.getSocketAddress().getAddress().getHostAddress())) {
```

Comment:
is `member.getSocketAddress().getAddress().getHostAddress()` null safe?

Reply:
The added member should have IP address attached as this method is executed on member add event. Hence the socket address object should have a value but if needed we can have a null check.